src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs

```diff
@@ -230,7 +230,20 @@ private ISymbol SuppressMessageAttribute
             }
         }
 
-        private void DecodeGlobalSuppressMessageAttributes()
+        private ISymbol? UnconditionalSuppressMessageAttribute
+        {
+            get
+            {
+                if (_lazyUnconditionalSuppressMessageAttribute == null)
+                {
+                    _lazyUnconditionalSuppressMessageAttribute = _compilation.GetTypeByMetadataName("System.Diagnostics.CodeAnalysis.UnconditionalSuppressMessageAttribute");
```

Comment:
>GetTypeByMetadataName [](start = 78, length = 21)

I think this API can return null, so using null as a sentinel for ```_lazyUnconditionalSuppressMessageAttribute``` is probably not optimal.

Reply:
This should be implemented according to the design in https://github.com/dotnet/roslyn/pull/51571

Reply:
> This should be implemented according to the design in [https://github.com/dotnet/roslyn/pull/51571](https://github.com/dotnet/roslyn/pull/51571)

Do you think it isn't?


---
In reply to: [595559328](https://github.com/dotnet/roslyn/pull/51792#discussion_r595559328) [](ancestors = 595559328)

Reply:
It seems to only apply to `GetWellKnownType`?

Reply:
> It seems to only apply to GetWellKnownType?

Are you judging by PR description? I think it is imprecise, to narrow. Check out the tests in the PR.

---
In reply to: [595589343](https://github.com/dotnet/roslyn/pull/51792#discussion_r595589343) [](ancestors = 595589343)

Reply:
In this case, I'd be looking for a test like the following:

* Assembly TA declares the attribute as `internal`
* Assembly TB declared the attribute as `public`
* Assembly TC references TA and TB, and uses the attribute

Verify that the suppression, which resolves to the attribute declared in TB, works properly.

Reply:
> Verify that the suppression, which resolves to the attribute declared in TB, works properly.

I don't think GetTypeByMetadataName cares about accessibility. In any case, this component already relies on GetTypeByMetadataName API for System.Diagnostics.CodeAnalysis.SuppressMessageAttribute. If you think lookup should be performed differently, I would say both use-sites should be adjusted and that is probably out of scope for this PR.

---
In reply to: [595600646](https://github.com/dotnet/roslyn/pull/51792#discussion_r595600646) [](ancestors = 595600646)

Reply:
SuppressMessageAttribute was added in .NET 2.0, so there wouldn't be as much of a need for someone to declare their own copy. The new attribute is much more likely to be impacted.

Reply:
> The new attribute is much more likely to be impacted.

Perhaps. 
1) Rather than trying to infer if the behavior matches your expectations from tests, it would probably be better if you could specify the behavior you want first.
2) I think we might still want to use the same lookup rules for both attributes. 
3) Andy is probably not the right person to implement new lookup rules (if we really want new lookup rules). Should someone from Analyzers team take over this PR? Or can this be done as a follow-up?

---
In reply to: [595615767](https://github.com/dotnet/roslyn/pull/51792#discussion_r595615767) [](ancestors = 595615767)

Reply:
For example, you might want to switch to the way compilers recognize attributes - fully qualified name of the type + signature of the constructor used. The lookup of a "blessed" type is not involved and that avoids the need to deal with all kinds of possible lookup ambiguities. An attribute with the right name and signature is applied, that is the "right" attribute and it should be respected. 

---
In reply to: [595619781](https://github.com/dotnet/roslyn/pull/51792#discussion_r595619781) [](ancestors = 595619781,595615767)

Reply:
Yeah, @AlekseyTs has a good point here -- I was actually surprised that this code looks for a particular type definition for attributes. In the compiler attributes are often accepted as long as they have the right name and signature, regardless of where they are defined.